### ISSUE TITLE
[BUG] compound matching with grain in nodegroup fails when used with file_tree ext_pillar

### ISSUE THREAD

**Issue by wchao** [saltstack/salt#58094]
**Description**
I use the file_tree ext_pillar to supply files to a group of minions using external pillars. I target the minions with a nodegroup definition in /etc/salt/master that uses a grain match. Only one of the minions is able to access the files as pillars. The rest of the minions report that the pillar does not exist.

**Setup**
/etc/salt/master:
ext_pillar:
  - file_tree:
      root_dir: /srv/pillar
      keep_newline: True
nodegroups:
  nodegroup_a: 'G@mygrain:abc*'

In /srv/pillar, I have /srv/pillar/nodegroups/nodegroup_a/files/readme1.txt.

I have minions minion1, minion2, minion3, and they have mygrain: abc1, mygrain: abc2, mygrain:abc3, respectively, in /etc/salt/grains.

/srv/salt/top.sls:
base:
  'minion1':
    - test
  'minion2':
    - test
  'minion3':
    - test

/srv/salt/test/init.sls:
/tmp/readme1.txt:
  file.managed:
    - contents_pillar: files:readme1.txt

**Steps to Reproduce the behavior**
salt minion1 state.apply
salt minion2 state.apply
salt minion3 state.apply

**Expected behavior**
I expect /tmp/readme1.txt to show up on minion1, minion2, minion3. Instead, it will show up on minion1, and minion2 and minion3 will report that Pillar files:readme1.txt does not exist.

If I replace nodegroup_a: 'G@mygrain:abc*' with nodegroup_a: 'L@minion1,minion2,minion3' in /etc/salt/master, everything works fine and all three minions are able to access the pillar files:readme1.txt.

**Screenshots**

**Versions Report**
<details><summary>salt --versions-report</summary>
(Provided by running salt --versions-report. Please also mention any differences in master/minion versions.) 

```
Salt Version:
           Salt: 3001

Dependency Versions:
           cffi: Not Installed
       cherrypy: Not Installed
       dateutil: 2.7.3
      docker-py: Not Installed
          gitdb: 2.0.5
      gitpython: 2.1.11
         Jinja2: 2.10
        libgit2: Not Installed
       M2Crypto: Not Installed
           Mako: Not Installed
   msgpack-pure: Not Installed
 msgpack-python: 0.5.6
   mysql-python: Not Installed
      pycparser: Not Installed
       pycrypto: Not Installed
   pycryptodome: 3.6.1
         pygit2: Not Installed
         Python: 3.7.3 (default, Dec 20 2019, 18:57:59)
   python-gnupg: Not Installed
         PyYAML: 3.13
          PyZMQ: 17.1.2
          smmap: 2.0.5
        timelib: Not Installed
        Tornado: 4.5.3
            ZMQ: 4.3.1

System Versions:
           dist: debian 10 buster
         locale: UTF-8
        machine: x86_64
        release: 4.19.0-9-amd64
         system: Linux
        version: Debian GNU/Linux 10 buster
```
</details>

**Additional context**

**Comment by corywright** [comment]
I believe I'm also being affected by this bug after upgrading from 2019.2.4 to 3001.3.

I use the file_tree ext pillar to track backend servers that should be excluded from a load balancer pool. 

```sls
ext_pillar:
  - file_tree:
      root_dir: /srv/pillar/file_tree
```
Nodegroup:
```yaml
nodegroups:
  openstack-lb:      'G@nodetype:openstack-lb'
```
Whenever I make changes to the files in the file_tree directory I always run the following commands before highstates simply to verify the pillar data is what I expect:

```bash
salt -N openstack-lb saltutil.refresh_pillar
salt -N openstack-lb pillar.get <name of file_tree pillar variable>
```

Sometimes, not always, but what appears to be around 20%-30% of the time, the minions do not receive any data at all from the file_tree pillar and the `pillar.get` call returns nothing at all. This should never happen as there are always lines in the files in our file tree root.

What's strange is, when this happens then each successive run of either `pillar.get` or `state.highstate` will be correct, then empty again. Meaning, that on one run it'll be empty, and if I immediately run the same command again, the pillar data will appear. If I then run it again, it'll be empty again, and then it'll show again. I've tested this many, many times and as bizarre as it is, this is what it's doing. We only have a single salt-master and this has all been working flawlessly for years until our recent upgrade to 3001.3.

The master and minions are all running the same OS (Ubuntu 18.04.5) and the same version of SaltStack. We previously were running 2019.2.4+ds-1 on Python 2 from the official apt repo, however we recently upgraded to 3001.3+ds-1 on Python 3 from the official apt repo.

This is a high severity bug for us -- the first time I noticed this happening it caused our failover database server to be added to the load balancer pool. We rely on pillar data being correct.

```
$ salt --versions-report
Salt Version:
           Salt: 3001.3
 
Dependency Versions:
           cffi: 1.11.5
       cherrypy: unknown
       dateutil: 2.7.3
      docker-py: Not Installed
          gitdb: 2.0.3
      gitpython: 3.0.7
         Jinja2: 2.10
        libgit2: Not Installed
       M2Crypto: Not Installed
           Mako: 1.0.7
   msgpack-pure: Not Installed
 msgpack-python: 0.6.2
   mysql-python: Not Installed
      pycparser: 2.18
       pycrypto: 2.6.1
   pycryptodome: 3.4.7
         pygit2: Not Installed
         Python: 3.6.9 (default, Jul 17 2020, 12:50:27)
   python-gnupg: 0.4.1
         PyYAML: 5.3.1
          PyZMQ: 17.1.2
          smmap: 2.0.3
        timelib: Not Installed
        Tornado: 4.5.3
            ZMQ: 4.2.5
 
System Versions:
           dist: ubuntu 18.04 bionic
         locale: UTF-8
        machine: x86_64
        release: 4.15.0-118-generic
         system: Linux
        version: Ubuntu 18.04 bionic
```

**Comment by corywright** [comment]
The #58611 issue also seems to be very similar to what I'm experiencing, although with a different ext_pillar.

**Comment by wchao** [comment]
Yes, I found the simple list matcher (L@) works fine, though obviously it's less than ideal (and potentially even unusable) for groups that dynamically change due to minions that have changing properties. In my case, the list of minions is fairly static, but changes once every few months, so I can live with L@, though I'd prefer matching on grains (G@). It's not a big enough problem for me that it's worth spending hours figuring out what's wrong in the code, but maybe that will change down the road....
